### PR TITLE
Laboratory work #3, Golovleva Yekaterina - 20FPL3

### PR DESCRIPTION
**Pull request by esgolovleva**

### CODE REVIEW

File: lab_3/main.py [hunk 1]
```diff
@@ -20,7 +21,37 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+        for word in tokens:
+            for letter in word:
+                letter.replace('ö', 'oe')
+                letter.replace('ü', 'ue')
+                letter.replace('ä', 'ae')
+                letter.replace('ß', 'ss')
+        list_of_sentences = []
+        for token in tokens:
+            list_of_sentences.append(tuple(['_'] + list(token) + ['_']))
+        if list_of_sentences == [] or len(sentence) == 1:
```
Comment (WhiteJaeger) [inline lab_3/main.py:50]: `list_of_sentences == []` can be done without explicit comparison.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_3/main.py (hunk 1 above)
Comment (esgolovleva) [reply lab_3/main.py:50]: I have replaced this part of string:
list_of_sentences == [] 


with this one:
if not list_of_sentences

File: lab_3/main.py [hunk 2]
```diff
@@ -20,7 +22,37 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+    result = []
+
+    invaluable_trash = ['`', '~', '!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '-', '+',
+                        '=', '{', '[', ']', '}', '|', '\\', ':', ';', '"', "'", '<', ',', '>',
+                        '.', '?', '/', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0']
+
+    for sentence in sentences:
+        sentence = sentence.lower()
+        for symbols in invaluable_trash:
+            sentence = sentence.replace(symbols, '')
```
Comment (WhiteJaeger) [inline lab_3/main.py:40]: Maybe replace them once in the whole text? Before these loops?

File: lab_3/main.py [hunk 3]
```diff
@@ -20,7 +22,37 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+        tokens = sentence.split()
+        for word in tokens:
+            for letter in word:
+                letter.replace('ö', 'oe')
+                letter.replace('ü', 'ue')
+                letter.replace('ä', 'ae')
+                letter.replace('ß', 'ss')
```
Comment (WhiteJaeger) [inline lab_3/main.py:47]: Maybe replace them once in the whole text? Before these loops?

File: lab_3/main.py [hunk 4]
```diff
@@ -222,7 +355,27 @@ def save(self, name: str) -> int:
...
+                    string_for_file += self.storage.get_letter_by_id(element)
+                freq_dict[string_for_file] = value
+
+        profile_as_dict['freq'] = freq_dict
+        profile_as_dict['n_words'] = self.n_words
+        profile_as_dict['name'] = self.language
+        with open(name, 'w', encoding="UTF-8") as file:
+            # added parameter "encoding" because of code style
+            json_string = json.dumps(profile_as_dict)
+            file.write(json_string)
```
Comment (WhiteJaeger) [inline lab_3/main.py:377]: There is a `.dump` method in the `json` package that combines these lines.

File: lab_3/main.py [hunk 5]
```diff
@@ -234,11 +387,34 @@ def open(self, file_name: str) -> int:
...
-        pass
+        if not isinstance(file_name, str):
+            return 1
+        with open(file_name, 'r', encoding="UTF-8") as file:
+            profile_dict = json.load(file)
+        self.language = profile_dict['name']
+        self.n_words = profile_dict['n_words']
+        freq_dict = profile_dict['freq']
+
+        size_of_gram = ''
```
Comment (WhiteJaeger) [inline lab_3/main.py:398]: Why do you define this as a string and later compare it with `len`?

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_3/main.py (hunk 4 above)
Comment (esgolovleva) [reply lab_3/main.py]: Ok, I fixed this

File: lab_3/main.py (hunk 5 above)
Comment (esgolovleva) [reply lab_3/main.py]: Ok, I fixed this.
Now the parameter size_of_gram = 0

**Pull request closed without merging**